// @steamship/client/build/main/lib/key_value_store.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KeyValueStore = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _util = require("util");
var _file = require("./file.js");
var _steamship_error = require("./steamship_error.js");
var _tag = require("./tag.js");
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
var KV_STORE_MARKER = '__init__';
var KeyValueStore = /*#__PURE__*/function () {
  /* A simple key value store implemented in Steamship.
       Instances of the KeyValueStore are identified by its  `namespace`.
        This store_identifier corresponds to a File that will be created with a special tag identifying it.
         Entries of the KeyValueStore are saved as `Tag` objects with:
      * Kind = "KeyValueStore"
      * Name = the key of the (kv) pair
      * Value = a dict set to the value
       Note that the value is always saved as a dict object. To save a string or int, wrap it in a dict.
         WARNING:
       This is essentially a clever hack atop Steamship's tag system to provide mutable key-value storage. It is in the
      steamship.utils package because it's proven useful once or twice. But in general, if you find yourself heavily
      relying upon it, consider reaching out to us at hello@steamship.com to let us know, and we'll up-prioritize
      adding a proper key-value API.
      */

  function KeyValueStore(client) {
    var storeIdentifier = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'KeyValueStore';
    (0, _classCallCheck2["default"])(this, KeyValueStore);
    (0, _defineProperty2["default"])(this, "client", void 0);
    (0, _defineProperty2["default"])(this, "storeIdentifier", void 0);
    this.client = client;
    this.storeIdentifier = "kv-store-".concat(storeIdentifier);
  }
  (0, _createClass2["default"])(KeyValueStore, [{
    key: "_getFile",
    value: function () {
      var _getFile2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var orCreate,
          statusFilesP,
          dummyContent,
          textEncoder,
          utf8,
          buffer,
          createP,
          _args = arguments;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              orCreate = _args.length > 0 && _args[0] !== undefined ? _args[0] : false;
              _context.next = 3;
              return _file.File.query(this.client, "filetag and kind \"".concat(this.storeIdentifier, "\""));
            case 3:
              statusFilesP = _context.sent;
              if (!(statusFilesP && statusFilesP.output && statusFilesP.output.files && statusFilesP.output.files.length > 0)) {
                _context.next = 6;
                break;
              }
              return _context.abrupt("return", statusFilesP.output.files[0]);
            case 6:
              if (orCreate) {
                _context.next = 8;
                break;
              }
              return _context.abrupt("return", undefined);
            case 8:
              dummyContent = '';
              textEncoder = new _util.TextEncoder();
              utf8 = new Uint8Array(dummyContent.length);
              textEncoder.encodeInto(dummyContent, utf8);
              buffer = Buffer.from(utf8);
              _context.next = 15;
              return _file.File.create(this.client, {
                content: buffer,
                tags: [{
                  kind: this.storeIdentifier,
                  name: KV_STORE_MARKER
                }]
              });
            case 15:
              createP = _context.sent;
              return _context.abrupt("return", createP.output);
            case 17:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function _getFile() {
        return _getFile2.apply(this, arguments);
      }
      return _getFile;
    }()
  }, {
    key: "get",
    value: function () {
      var _get = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(key) {
        var file, _iterator, _step, tag;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return this._getFile();
            case 2:
              file = _context2.sent;
              if (file) {
                _context2.next = 5;
                break;
              }
              return _context2.abrupt("return", undefined);
            case 5:
              if (file.tags) {
                _context2.next = 7;
                break;
              }
              return _context2.abrupt("return", undefined);
            case 7:
              _iterator = _createForOfIteratorHelper(file.tags);
              _context2.prev = 8;
              _iterator.s();
            case 10:
              if ((_step = _iterator.n()).done) {
                _context2.next = 17;
                break;
              }
              tag = _step.value;
              if (!(tag.kind == this.storeIdentifier && tag.name == key)) {
                _context2.next = 15;
                break;
              }
              console.log('got tag', tag);
              return _context2.abrupt("return", tag.value);
            case 15:
              _context2.next = 10;
              break;
            case 17:
              _context2.next = 22;
              break;
            case 19:
              _context2.prev = 19;
              _context2.t0 = _context2["catch"](8);
              _iterator.e(_context2.t0);
            case 22:
              _context2.prev = 22;
              _iterator.f();
              return _context2.finish(22);
            case 25:
              return _context2.abrupt("return", undefined);
            case 26:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this, [[8, 19, 22, 25]]);
      }));
      function get(_x) {
        return _get.apply(this, arguments);
      }
      return get;
    }()
  }, {
    key: "delete",
    value: function () {
      var _delete2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(key) {
        var file, _iterator2, _step2, tag;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              _context3.next = 2;
              return this._getFile();
            case 2:
              file = _context3.sent;
              if (file) {
                _context3.next = 5;
                break;
              }
              return _context3.abrupt("return", false);
            case 5:
              if (file.tags) {
                _context3.next = 7;
                break;
              }
              return _context3.abrupt("return", false);
            case 7:
              _iterator2 = _createForOfIteratorHelper(file.tags);
              _context3.prev = 8;
              _iterator2.s();
            case 10:
              if ((_step2 = _iterator2.n()).done) {
                _context3.next = 18;
                break;
              }
              tag = _step2.value;
              if (!(tag.kind == this.storeIdentifier && tag.name == key)) {
                _context3.next = 16;
                break;
              }
              _context3.next = 15;
              return tag["delete"]();
            case 15:
              return _context3.abrupt("return", true);
            case 16:
              _context3.next = 10;
              break;
            case 18:
              _context3.next = 23;
              break;
            case 20:
              _context3.prev = 20;
              _context3.t0 = _context3["catch"](8);
              _iterator2.e(_context3.t0);
            case 23:
              _context3.prev = 23;
              _iterator2.f();
              return _context3.finish(23);
            case 26:
              return _context3.abrupt("return", false);
            case 27:
            case "end":
              return _context3.stop();
          }
        }, _callee3, this, [[8, 20, 23, 26]]);
      }));
      function _delete(_x2) {
        return _delete2.apply(this, arguments);
      }
      return _delete;
    }()
  }, {
    key: "set",
    value: function () {
      var _set = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(key, value) {
        var file;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              _context4.next = 2;
              return this["delete"](key);
            case 2:
              _context4.next = 4;
              return this._getFile(true);
            case 4:
              file = _context4.sent;
              if (file) {
                _context4.next = 7;
                break;
              }
              throw new _steamship_error.SteamshipError({
                statusMessage: 'Unable to create file for tag.'
              });
            case 7:
              return _context4.abrupt("return", _tag.Tag.create(this.client, {
                fileId: file.id,
                kind: this.storeIdentifier,
                name: key,
                value: value
              }));
            case 8:
            case "end":
              return _context4.stop();
          }
        }, _callee4, this);
      }));
      function set(_x3, _x4) {
        return _set.apply(this, arguments);
      }
      return set;
    }() // """Return all key-value entries as a list of (key, value) tuples.
    //
    // If `filter_keys` is provided, only returns keys within that list."""
  }, {
    key: "items",
    value: function () {
      var _items = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(filterKeys) {
        var file, ret, _iterator3, _step3, tag;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) switch (_context5.prev = _context5.next) {
            case 0:
              _context5.next = 2;
              return this._getFile();
            case 2:
              file = _context5.sent;
              if (!(!file || !file.tags)) {
                _context5.next = 5;
                break;
              }
              return _context5.abrupt("return", []);
            case 5:
              ret = [];
              _iterator3 = _createForOfIteratorHelper(file.tags);
              try {
                for (_iterator3.s(); !(_step3 = _iterator3.n()).done;) {
                  tag = _step3.value;
                  if (tag.kind == this.storeIdentifier && tag.name != KV_STORE_MARKER && (!filterKeys || tag.name && filterKeys.includes(tag.name))) {
                    if (tag.name) {
                      ret.push([tag.name, tag.value]);
                    }
                  }
                }
              } catch (err) {
                _iterator3.e(err);
              } finally {
                _iterator3.f();
              }
              return _context5.abrupt("return", ret);
            case 9:
            case "end":
              return _context5.stop();
          }
        }, _callee5, this);
      }));
      function items(_x5) {
        return _items.apply(this, arguments);
      }
      return items;
    }()
  }, {
    key: "reset",
    value: function () {
      var _reset = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6() {
        var file;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) switch (_context6.prev = _context6.next) {
            case 0:
              _context6.next = 2;
              return this._getFile();
            case 2:
              file = _context6.sent;
              if (!file) {
                _context6.next = 6;
                break;
              }
              _context6.next = 6;
              return file["delete"]();
            case 6:
            case "end":
              return _context6.stop();
          }
        }, _callee6, this);
      }));
      function reset() {
        return _reset.apply(this, arguments);
      }
      return reset;
    }()
  }]);
  return KeyValueStore;
}();
exports.KeyValueStore = KeyValueStore;
