// @steamship/client/build/module/lib/api_base.js
function _regeneratorRuntime() { "use strict"; /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/facebook/regenerator/blob/main/LICENSE */ _regeneratorRuntime = function _regeneratorRuntime() { return exports; }; var exports = {}, Op = Object.prototype, hasOwn = Op.hasOwnProperty, defineProperty = Object.defineProperty || function (obj, key, desc) { obj[key] = desc.value; }, $Symbol = "function" == typeof Symbol ? Symbol : {}, iteratorSymbol = $Symbol.iterator || "@@iterator", asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator", toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag"; function define(obj, key, value) { return Object.defineProperty(obj, key, { value: value, enumerable: !0, configurable: !0, writable: !0 }), obj[key]; } try { define({}, ""); } catch (err) { define = function define(obj, key, value) { return obj[key] = value; }; } function wrap(innerFn, outerFn, self, tryLocsList) { var protoGenerator = outerFn && outerFn.prototype instanceof Generator ? outerFn : Generator, generator = Object.create(protoGenerator.prototype), context = new Context(tryLocsList || []); return defineProperty(generator, "_invoke", { value: makeInvokeMethod(innerFn, self, context) }), generator; } function tryCatch(fn, obj, arg) { try { return { type: "normal", arg: fn.call(obj, arg) }; } catch (err) { return { type: "throw", arg: err }; } } exports.wrap = wrap; var ContinueSentinel = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} var IteratorPrototype = {}; define(IteratorPrototype, iteratorSymbol, function () { return this; }); var getProto = Object.getPrototypeOf, NativeIteratorPrototype = getProto && getProto(getProto(values([]))); NativeIteratorPrototype && NativeIteratorPrototype !== Op && hasOwn.call(NativeIteratorPrototype, iteratorSymbol) && (IteratorPrototype = NativeIteratorPrototype); var Gp = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(IteratorPrototype); function defineIteratorMethods(prototype) { ["next", "throw", "return"].forEach(function (method) { define(prototype, method, function (arg) { return this._invoke(method, arg); }); }); } function AsyncIterator(generator, PromiseImpl) { function invoke(method, arg, resolve, reject) { var record = tryCatch(generator[method], generator, arg); if ("throw" !== record.type) { var result = record.arg, value = result.value; return value && "object" == typeof value && hasOwn.call(value, "__await") ? PromiseImpl.resolve(value.__await).then(function (value) { invoke("next", value, resolve, reject); }, function (err) { invoke("throw", err, resolve, reject); }) : PromiseImpl.resolve(value).then(function (unwrapped) { result.value = unwrapped, resolve(result); }, function (error) { return invoke("throw", error, resolve, reject); }); } reject(record.arg); } var previousPromise; defineProperty(this, "_invoke", { value: function value(method, arg) { function callInvokeWithMethodAndArg() { return new PromiseImpl(function (resolve, reject) { invoke(method, arg, resolve, reject); }); } return previousPromise = previousPromise ? previousPromise.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg(); } }); } function makeInvokeMethod(innerFn, self, context) { var state = "suspendedStart"; return function (method, arg) { if ("executing" === state) throw new Error("Generator is already running"); if ("completed" === state) { if ("throw" === method) throw arg; return doneResult(); } for (context.method = method, context.arg = arg;;) { var delegate = context.delegate; if (delegate) { var delegateResult = maybeInvokeDelegate(delegate, context); if (delegateResult) { if (delegateResult === ContinueSentinel) continue; return delegateResult; } } if ("next" === context.method) context.sent = context._sent = context.arg;else if ("throw" === context.method) { if ("suspendedStart" === state) throw state = "completed", context.arg; context.dispatchException(context.arg); } else "return" === context.method && context.abrupt("return", context.arg); state = "executing"; var record = tryCatch(innerFn, self, context); if ("normal" === record.type) { if (state = context.done ? "completed" : "suspendedYield", record.arg === ContinueSentinel) continue; return { value: record.arg, done: context.done }; } "throw" === record.type && (state = "completed", context.method = "throw", context.arg = record.arg); } }; } function maybeInvokeDelegate(delegate, context) { var methodName = context.method, method = delegate.iterator[methodName]; if (undefined === method) return context.delegate = null, "throw" === methodName && delegate.iterator["return"] && (context.method = "return", context.arg = undefined, maybeInvokeDelegate(delegate, context), "throw" === context.method) || "return" !== methodName && (context.method = "throw", context.arg = new TypeError("The iterator does not provide a '" + methodName + "' method")), ContinueSentinel; var record = tryCatch(method, delegate.iterator, context.arg); if ("throw" === record.type) return context.method = "throw", context.arg = record.arg, context.delegate = null, ContinueSentinel; var info = record.arg; return info ? info.done ? (context[delegate.resultName] = info.value, context.next = delegate.nextLoc, "return" !== context.method && (context.method = "next", context.arg = undefined), context.delegate = null, ContinueSentinel) : info : (context.method = "throw", context.arg = new TypeError("iterator result is not an object"), context.delegate = null, ContinueSentinel); } function pushTryEntry(locs) { var entry = { tryLoc: locs[0] }; 1 in locs && (entry.catchLoc = locs[1]), 2 in locs && (entry.finallyLoc = locs[2], entry.afterLoc = locs[3]), this.tryEntries.push(entry); } function resetTryEntry(entry) { var record = entry.completion || {}; record.type = "normal", delete record.arg, entry.completion = record; } function Context(tryLocsList) { this.tryEntries = [{ tryLoc: "root" }], tryLocsList.forEach(pushTryEntry, this), this.reset(!0); } function values(iterable) { if (iterable) { var iteratorMethod = iterable[iteratorSymbol]; if (iteratorMethod) return iteratorMethod.call(iterable); if ("function" == typeof iterable.next) return iterable; if (!isNaN(iterable.length)) { var i = -1, next = function next() { for (; ++i < iterable.length;) if (hasOwn.call(iterable, i)) return next.value = iterable[i], next.done = !1, next; return next.value = undefined, next.done = !0, next; }; return next.next = next; } } return { next: doneResult }; } function doneResult() { return { value: undefined, done: !0 }; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, defineProperty(Gp, "constructor", { value: GeneratorFunctionPrototype, configurable: !0 }), defineProperty(GeneratorFunctionPrototype, "constructor", { value: GeneratorFunction, configurable: !0 }), GeneratorFunction.displayName = define(GeneratorFunctionPrototype, toStringTagSymbol, "GeneratorFunction"), exports.isGeneratorFunction = function (genFun) { var ctor = "function" == typeof genFun && genFun.constructor; return !!ctor && (ctor === GeneratorFunction || "GeneratorFunction" === (ctor.displayName || ctor.name)); }, exports.mark = function (genFun) { return Object.setPrototypeOf ? Object.setPrototypeOf(genFun, GeneratorFunctionPrototype) : (genFun.__proto__ = GeneratorFunctionPrototype, define(genFun, toStringTagSymbol, "GeneratorFunction")), genFun.prototype = Object.create(Gp), genFun; }, exports.awrap = function (arg) { return { __await: arg }; }, defineIteratorMethods(AsyncIterator.prototype), define(AsyncIterator.prototype, asyncIteratorSymbol, function () { return this; }), exports.AsyncIterator = AsyncIterator, exports.async = function (innerFn, outerFn, self, tryLocsList, PromiseImpl) { void 0 === PromiseImpl && (PromiseImpl = Promise); var iter = new AsyncIterator(wrap(innerFn, outerFn, self, tryLocsList), PromiseImpl); return exports.isGeneratorFunction(outerFn) ? iter : iter.next().then(function (result) { return result.done ? result.value : iter.next(); }); }, defineIteratorMethods(Gp), define(Gp, toStringTagSymbol, "Generator"), define(Gp, iteratorSymbol, function () { return this; }), define(Gp, "toString", function () { return "[object Generator]"; }), exports.keys = function (val) { var object = Object(val), keys = []; for (var key in object) keys.push(key); return keys.reverse(), function next() { for (; keys.length;) { var key = keys.pop(); if (key in object) return next.value = key, next.done = !1, next; } return next.done = !0, next; }; }, exports.values = values, Context.prototype = { constructor: Context, reset: function reset(skipTempReset) { if (this.prev = 0, this.next = 0, this.sent = this._sent = undefined, this.done = !1, this.delegate = null, this.method = "next", this.arg = undefined, this.tryEntries.forEach(resetTryEntry), !skipTempReset) for (var name in this) "t" === name.charAt(0) && hasOwn.call(this, name) && !isNaN(+name.slice(1)) && (this[name] = undefined); }, stop: function stop() { this.done = !0; var rootRecord = this.tryEntries[0].completion; if ("throw" === rootRecord.type) throw rootRecord.arg; return this.rval; }, dispatchException: function dispatchException(exception) { if (this.done) throw exception; var context = this; function handle(loc, caught) { return record.type = "throw", record.arg = exception, context.next = loc, caught && (context.method = "next", context.arg = undefined), !!caught; } for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i], record = entry.completion; if ("root" === entry.tryLoc) return handle("end"); if (entry.tryLoc <= this.prev) { var hasCatch = hasOwn.call(entry, "catchLoc"), hasFinally = hasOwn.call(entry, "finallyLoc"); if (hasCatch && hasFinally) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } else if (hasCatch) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); } else { if (!hasFinally) throw new Error("try statement without catch or finally"); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } } } }, abrupt: function abrupt(type, arg) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc <= this.prev && hasOwn.call(entry, "finallyLoc") && this.prev < entry.finallyLoc) { var finallyEntry = entry; break; } } finallyEntry && ("break" === type || "continue" === type) && finallyEntry.tryLoc <= arg && arg <= finallyEntry.finallyLoc && (finallyEntry = null); var record = finallyEntry ? finallyEntry.completion : {}; return record.type = type, record.arg = arg, finallyEntry ? (this.method = "next", this.next = finallyEntry.finallyLoc, ContinueSentinel) : this.complete(record); }, complete: function complete(record, afterLoc) { if ("throw" === record.type) throw record.arg; return "break" === record.type || "continue" === record.type ? this.next = record.arg : "return" === record.type ? (this.rval = this.arg = record.arg, this.method = "return", this.next = "end") : "normal" === record.type && afterLoc && (this.next = afterLoc), ContinueSentinel; }, finish: function finish(finallyLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.finallyLoc === finallyLoc) return this.complete(entry.completion, entry.afterLoc), resetTryEntry(entry), ContinueSentinel; } }, "catch": function _catch(tryLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc === tryLoc) { var record = entry.completion; if ("throw" === record.type) { var thrown = record.arg; resetTryEntry(entry); } return thrown; } } throw new Error("illegal catch attempt"); }, delegateYield: function delegateYield(iterable, resultName, nextLoc) { return this.delegate = { iterator: values(iterable), resultName: resultName, nextLoc: nextLoc }, "next" === this.method && (this.arg = undefined), ContinueSentinel; } }, exports; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }
function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }
import axios from 'axios';
import getLogger from './log.js';
import { loadConfiguration } from './shared/Configuration.js';
import { SteamshipError } from './steamship_error.js';
import { Task } from './task.js';
import { TaskState } from './types/base.js';
import { isNode } from './utils.js';
function doFetch(_x, _x2) {
  return _doFetch.apply(this, arguments);
} // async function doFetch(url: any, opts: any): Promise<AxiosResponse> {
//   return new Promise<AxiosResponse>((resolve, reject) => {
//     axios({
//       url,
//       headers: opts.headers,
//       data: opts.body,
//       method: opts.method,
//     }).then(resolve)
//       .catch(
//         (err) => {
//           console.log(err);
//           reject(err);
//         }
//       )
//   })
// }
function _doFetch() {
  _doFetch = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee8(url, opts) {
    return _regeneratorRuntime().wrap(function _callee8$(_context8) {
      while (1) switch (_context8.prev = _context8.next) {
        case 0:
          return _context8.abrupt("return", axios({
            url: url,
            headers: opts.headers,
            data: opts.body,
            method: opts.method,
            responseType: opts.responseType
          }));
        case 1:
        case "end":
          return _context8.stop();
      }
    }, _callee8);
  }));
  return _doFetch.apply(this, arguments);
}
var log = getLogger('Steamship:ApiBase');
var MAX_BODY_LENGTH = 100000 * 1000;
var _SUBDOMAIN_REWRITING_ENDPOINTS = ['//steamship.run', '//apps.staging.steamship.com'];
var _USE_SUBDOMAIN_REWRITING = function _USE_SUBDOMAIN_REWRITING(base) {
  for (var _i = 0, _SUBDOMAIN_REWRITING_ = _SUBDOMAIN_REWRITING_ENDPOINTS; _i < _SUBDOMAIN_REWRITING_.length; _i++) {
    var s = _SUBDOMAIN_REWRITING_[_i];
    if (base.includes(s)) {
      return true;
    }
  }
  return false;
};

/* Should be a FormData object */
function addMultiparts(formData, path, value) {
  /* Decode any object into a series of HTTP Multi-part segments that Vapor will consume.
    https://github.com/vapor/multipart-kit
     When sending a JSON object in a MultiPart request, Vapor wishes to see multi part segments as follows:
    single_key
    array_key[idx]
    obj_key[prop]
     So a File with a list of one tag with kind=Foo would be transmitted as setting the part:
    [tags][0][kind]
  */
  var type = typeof value;
  if (type == 'string' || type == 'boolean' || type == 'number') {
    formData.append(path, value);
  } else if (Array.isArray(value)) {
    value.forEach(function (subValue, index) {
      addMultiparts(formData, path + "[" + index + "]", subValue);
    });
  } else {
    for (var _key in value) {
      addMultiparts(formData, path + "[" + _key + "]", value[_key]);
    }
  }
}
export var ApiBase = /*#__PURE__*/function () {
  function ApiBase(params) {
    var _this = this;
    _classCallCheck(this, ApiBase);
    _defineProperty(this, "config", void 0);
    this.config = loadConfiguration(params).then(function (config) {
      if (config.apiKey) {
        return _this._switchConfigWorkspace({
          workspaceHandle: params == null ? void 0 : params.workspace,
          failIfWorkspaceExists: (params == null ? void 0 : params.failIfWorkspaceExists) === true,
          config: config
        });
      } else {
        return config;
      }
    });
  }
  _createClass(ApiBase, [{
    key: "switchWorkspace",
    value: function () {
      var _switchWorkspace = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(params) {
        var config;
        return _regeneratorRuntime().wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.t0 = this;
              _context.t1 = _extends;
              _context.t2 = {};
              _context.t3 = params;
              _context.next = 6;
              return this.config;
            case 6:
              _context.t4 = _context.sent;
              _context.t5 = {
                config: _context.t4
              };
              _context.t6 = (0, _context.t1)(_context.t2, _context.t3, _context.t5);
              _context.next = 11;
              return _context.t0._switchConfigWorkspace.call(_context.t0, _context.t6);
            case 11:
              config = _context.sent;
              this.config = Promise.resolve(config);
              log.info("Switched to workspace " + config.workspaceHandle + "/" + config.workspaceId);
            case 14:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function switchWorkspace(_x3) {
        return _switchWorkspace.apply(this, arguments);
      }
      return switchWorkspace;
    }()
    /*
      Switches this client to the requested workspace, possibly creating it. If all arguments are None, the client
      actively switches into the default workspace.
       - API calls are performed manually to not result in circular imports.
      - Note that the default workspace is technically not necessary for API usage; it will be assumed by the Engine
        in the absense of a Workspace ID or Handle being manually specified in request headers.
     */
  }, {
    key: "_switchConfigWorkspace",
    value: function () {
      var _switchConfigWorkspace2 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee2(params) {
        var workspace, p, oldConfig, activeConfig, returnId, returnHandle;
        return _regeneratorRuntime().wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              workspace = undefined;
              p = _extends({}, params);
              if (!p.workspaceHandle) {
                p.workspaceHandle = 'default';
              }
              if (p.failIfWorkspaceExists) {
                log.info("Creating workspace with handle: " + p.workspaceHandle + ".");
              } else {
                log.info("Creating/Fetching workspace with handle: " + p.workspaceHandle + ".");
              }

              // Zero out the workspace_handle on the config block in case we're being invoked from
              // `init` (otherwise we'll attempt to create the sapce IN that nonexistant workspace)
              oldConfig = params.config;
              activeConfig = _extends({}, oldConfig, {
                workspaceHandle: undefined,
                workspaceId: undefined
              });
              _context2.next = 8;
              return this.post('workspace/create', {
                handle: p.workspaceHandle,
                fetchIfExists: !(p.failIfWorkspaceExists === true)
              }, {
                responsePath: 'workspace'
              }, activeConfig);
            case 8:
              workspace = _context2.sent;
              if (!(!workspace || !workspace.output)) {
                _context2.next = 11;
                break;
              }
              throw new SteamshipError({
                statusMessage: 'Was unable to switch to new workspace: server returned empty Workspace.'
              });
            case 11:
              returnId = workspace.output.id;
              returnHandle = workspace.output.handle;
              if (!(!returnHandle || !returnId)) {
                _context2.next = 16;
                break;
              }
              log.error(workspace);
              throw new SteamshipError({
                statusMessage: JSON.stringify(workspace) + " Unable to switch to workspace " + p.workspaceHandle + " with failIfWorkspaceExists " + p.failIfWorkspaceExists + ": server returned empty ID and Handle."
              });
            case 16:
              return _context2.abrupt("return", _extends({}, oldConfig, {
                workspaceHandle: returnHandle,
                workspaceId: returnId
              }));
            case 17:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function _switchConfigWorkspace(_x4) {
        return _switchConfigWorkspace2.apply(this, arguments);
      }
      return _switchConfigWorkspace;
    }()
  }, {
    key: "_headers",
    value: function _headers(config, postConfig) {
      var ret = {
        'Content-Type': 'application/json',
        Authorization: "Bearer " + config.apiKey
      };
      if (postConfig != null && postConfig.workspaceId) {
        ret['X-Workspace-Id'] = postConfig == null ? void 0 : postConfig.workspaceId;
      } else if (config.workspaceId) {
        ret['X-Workspace-Id'] = config.workspaceId;
      } else if (config.workspaceHandle) {
        ret['X-Workspace-Handle'] = config.workspaceHandle;
      }
      if ((postConfig == null ? void 0 : postConfig.isPackageCall) === true) {
        if (postConfig != null && postConfig.packageOwner) {
          ret['X-Package-Owner-Handle'] = postConfig.packageOwner;
        }
        if (postConfig != null && postConfig.packageId) {
          ret['X-Package-Id'] = postConfig.packageId;
        }
        if (postConfig != null && postConfig.packageInstanceId) {
          ret['X-Package-Instance-Id'] = postConfig.packageInstanceId;
        }
      }
      if ((postConfig == null ? void 0 : postConfig.asBackgroundTask) === true) {
        // Will result in the engine persisting the inbound HTTP request as a Task for deferred
        // execution. The client will receive task information back instead of the synchronous API response.
        // That task can be polled for eventual completion.
        ret['X-Task-Background'] = 'true';
      }
      return ret;
    }
  }, {
    key: "_url",
    value: function _url(baseConfig, postConfig, operation) {
      var base = '';
      if (!((postConfig == null ? void 0 : postConfig.isPackageCall) === true)) {
        // It's an API call
        base = (postConfig == null ? void 0 : postConfig.apiBase) || baseConfig.apiBase;
      } else {
        // It's a package call
        if (!(postConfig != null && postConfig.packageOwner)) {
          throw new SteamshipError({
            statusCode: 'UserMissing',
            statusMessage: "Can not invoke an package endpoint without the package owner's user handle.",
            statusSuggestion: 'Provide the packageOwner option, or initialize your package with a lookup.'
          });
        }
        base = (postConfig == null ? void 0 : postConfig.appBase) || baseConfig.appBase;
        if (!base) {
          throw new SteamshipError({
            statusCode: 'EndpointMissing',
            statusMessage: 'Can not invoke an package endpoint without the App Base variable set.',
            statusSuggestion: 'This should automatically have a good default setting. Reach out to our Steamship support.'
          });
        }
        if (_USE_SUBDOMAIN_REWRITING(base)) {
          // Rewrite the base to be https://user.base
          var parts = base.split('//');
          if (parts.length < 2) {
            throw new SteamshipError({
              statusCode: 'EndpointInvalid',
              statusMessage: 'You package base did not appear to begin with a valid HTTP or HTTPS protocol.',
              statusSuggestion: "Make sure you've provided an app base such as https://steamship.run, with the protocol."
            });
          }
          // Now we pre-pend the app-base to the first part!
          parts[1] = (postConfig == null ? void 0 : postConfig.packageOwner) + "." + parts[1];
          base = parts.join('//');
        }
      }

      // Guard against a double // after the domain

      if (base && base[base.length - 1] == '/') {
        base = base.slice(0, base.length - 1);
      }
      if (operation && operation[0] == '/') {
        operation = operation.slice(1);
      }
      return base + "/" + operation;
    }
  }, {
    key: "post",
    value: function () {
      var _post = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee3(operation, payload, config, overrideConfig) {
        return _regeneratorRuntime().wrap(function _callee3$(_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              return _context3.abrupt("return", this.call('POST', operation, payload, config, overrideConfig));
            case 1:
            case "end":
              return _context3.stop();
          }
        }, _callee3, this);
      }));
      function post(_x5, _x6, _x7, _x8) {
        return _post.apply(this, arguments);
      }
      return post;
    }()
  }, {
    key: "get",
    value: function () {
      var _get = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee4(operation, payload, config) {
        return _regeneratorRuntime().wrap(function _callee4$(_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              return _context4.abrupt("return", this.call('GET', operation, payload, config));
            case 1:
            case "end":
              return _context4.stop();
          }
        }, _callee4, this);
      }));
      function get(_x9, _x10, _x11) {
        return _get.apply(this, arguments);
      }
      return get;
    }()
  }, {
    key: "_makeError",
    value: function () {
      var _makeError2 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee5(_ref) {
        var error, response, task, j;
        return _regeneratorRuntime().wrap(function _callee5$(_context5) {
          while (1) switch (_context5.prev = _context5.next) {
            case 0:
              error = _ref.error, response = _ref.response, task = _ref.task;
              if (!error) {
                _context5.next = 5;
                break;
              }
              return _context5.abrupt("return", error);
            case 5:
              if (!response) {
                _context5.next = 20;
                break;
              }
              _context5.prev = 6;
              _context5.next = 9;
              return response.data;
            case 9:
              j = _context5.sent;
              if (!j.status) {
                _context5.next = 12;
                break;
              }
              return _context5.abrupt("return", new SteamshipError({
                statusMessage: j.status.statusMessage,
                statusCode: j.status.statusCode,
                statusSuggestion: j.status.statusSuggestion
              }));
            case 12:
              _context5.next = 17;
              break;
            case 14:
              _context5.prev = 14;
              _context5.t0 = _context5["catch"](6);
              return _context5.abrupt("return", new SteamshipError({
                statusMessage: 'Bad response from server'
              }));
            case 17:
              return _context5.abrupt("return", new SteamshipError({
                statusMessage: 'Bad response from server'
              }));
            case 20:
              if (!task) {
                _context5.next = 22;
                break;
              }
              return _context5.abrupt("return", new SteamshipError({
                statusMessage: task.statusMessage,
                statusCode: task.statusCode,
                statusSuggestion: task.statusSuggestion
              }));
            case 22:
              return _context5.abrupt("return", new SteamshipError({
                statusMessage: 'Bad response'
              }));
            case 23:
            case "end":
              return _context5.stop();
          }
        }, _callee5, null, [[6, 14]]);
      }));
      function _makeError(_x12) {
        return _makeError2.apply(this, arguments);
      }
      return _makeError;
    }()
  }, {
    key: "_makeResponse",
    value: function () {
      var _makeResponse2 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee6(_ref2) {
        var _json, _json2;
        var response, rawResponse, objectConstructor, responsePath, err, json, taskParams, task, _task;
        return _regeneratorRuntime().wrap(function _callee6$(_context6) {
          while (1) switch (_context6.prev = _context6.next) {
            case 0:
              response = _ref2.response, rawResponse = _ref2.rawResponse, objectConstructor = _ref2.objectConstructor, responsePath = _ref2.responsePath;
              if (response) {
                _context6.next = 3;
                break;
              }
              throw new SteamshipError({
                statusMessage: 'No response.'
              });
            case 3:
              if (!(response.status != 200)) {
                _context6.next = 8;
                break;
              }
              _context6.next = 6;
              return this._makeError({
                response: response
              });
            case 6:
              err = _context6.sent;
              throw err;
            case 8:
              if (!rawResponse) {
                _context6.next = 10;
                break;
              }
              return _context6.abrupt("return", response);
            case 10:
              _context6.prev = 10;
              json = response.data; // Axios
              // json = JSON.parse((response as any).body); // Got
              // json = await response.json(); // for the fetch style // Fetch
              _context6.next = 19;
              break;
            case 14:
              _context6.prev = 14;
              _context6.t0 = _context6["catch"](10);
              _context6.next = 18;
              return this._makeError({
                error: _context6.t0
              });
            case 18:
              throw _context6.sent;
            case 19:
              if (!(typeof json == 'undefined' || json == null)) {
                _context6.next = 21;
                break;
              }
              throw new SteamshipError({
                statusMessage: 'Empty response.'
              });
            case 21:
              taskParams = (_json = json) == null ? void 0 : _json.status;
              if (!(typeof (taskParams == null ? void 0 : taskParams.state) != 'undefined')) {
                _context6.next = 30;
                break;
              }
              task = new Task(this, _extends({}, taskParams, {
                responsePath: responsePath,
                rawResponse: rawResponse,
                objectConstructor: objectConstructor
              }));
              if (!(((_task = task) == null ? void 0 : _task.state) == TaskState.failed)) {
                _context6.next = 28;
                break;
              }
              _context6.next = 27;
              return this._makeError({
                task: task
              });
            case 27:
              throw _context6.sent;
            case 28:
              _context6.next = 31;
              break;
            case 30:
              task = new Task(this, {
                responsePath: responsePath,
                rawResponse: rawResponse,
                objectConstructor: objectConstructor
              });
            case 31:
              task.setData((_json2 = json) == null ? void 0 : _json2.data);
              return _context6.abrupt("return", task);
            case 33:
            case "end":
              return _context6.stop();
          }
        }, _callee6, this, [[10, 14]]);
      }));
      function _makeResponse(_x13) {
        return _makeResponse2.apply(this, arguments);
      }
      return _makeResponse;
    }()
  }, {
    key: "call",
    value: function () {
      var _call = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee7(verb, operation, payload, config, overrideConfig) {
        var baseConfig, url, headers, reqConfig, finalPayload, contentType, FormDataNode, formData, pp, _i2, _Object$keys, _key2, value, _formData, _pp, _i3, _Object$keys2, _key3, _value, resp;
        return _regeneratorRuntime().wrap(function _callee7$(_context7) {
          while (1) switch (_context7.prev = _context7.next) {
            case 0:
              _context7.t0 = overrideConfig;
              if (_context7.t0) {
                _context7.next = 5;
                break;
              }
              _context7.next = 4;
              return this.config;
            case 4:
              _context7.t0 = _context7.sent;
            case 5:
              baseConfig = _context7.t0;
              if (baseConfig.apiKey) {
                _context7.next = 8;
                break;
              }
              throw new SteamshipError({
                statusCode: 'Authentication',
                statusMessage: 'API Key not found.',
                statusSuggestion: 'Please see docs.steamship.com for a variety of ways to set your API key.'
              });
            case 8:
              url = this._url(baseConfig, config, operation);
              headers = this._headers(baseConfig, config);
              reqConfig = {
                headers: headers,
                maxContentLength: MAX_BODY_LENGTH,
                maxBodyLength: MAX_BODY_LENGTH
              };
              finalPayload = undefined;
              if (!(verb == 'POST' && config != null && config.file)) {
                _context7.next = 35;
                break;
              }
              contentType = undefined;
              if (typeof config.file != 'string') {
                contentType = 'binary/octet-stream';
              }

              // Important so proper boundary can be set;
              delete reqConfig.headers['Content-Type'];
              if (!isNode()) {
                _context7.next = 28;
                break;
              }
              _context7.next = 19;
              return import('form-data');
            case 19:
              FormDataNode = _context7.sent;
              formData = new FormDataNode["default"]();
              formData.append('file', Buffer.from(config.file), {
                filename: config == null ? void 0 : config.filename,
                contentType: contentType
              });
              pp = payload;
              for (_i2 = 0, _Object$keys = Object.keys(pp); _i2 < _Object$keys.length; _i2++) {
                _key2 = _Object$keys[_i2];
                value = pp[_key2];
                addMultiparts(formData, _key2, value);
              }
              finalPayload = formData;

              // This only needs to happen on Node.
              // And the .getHeaders method is unavilabile in the browser.
              // NOTE: This is untested in the unit tests; it will show up as a failure in the browser.
              // TODO: We need to start running tests inside a browser runtime too.
              reqConfig.headers = _extends({}, reqConfig.headers, formData.getHeaders());
              _context7.next = 33;
              break;
            case 28:
              _formData = new FormData();
              /*
               * The config.file as any cast below is because FormData does not support Buffer.
               * In general, file as a Buffer should only be done from within the NodeJS environment,
               * not from within the browser. In the browser, Blob should be used instead.
               *
               * TODO: Figure out if there's a way to strongly type (or at least runtime check) this
               * so that we get back intelligent errors that are environment-dependent.
               */
              _formData.append('file', config.file, config == null ? void 0 : config.filename);
              _pp = payload;
              for (_i3 = 0, _Object$keys2 = Object.keys(_pp); _i3 < _Object$keys2.length; _i3++) {
                _key3 = _Object$keys2[_i3];
                _value = _pp[_key3];
                addMultiparts(_formData, _key3, _value);
              }
              finalPayload = _formData;
            case 33:
              _context7.next = 36;
              break;
            case 35:
              if (typeof payload == 'object') {
                finalPayload = JSON.stringify(payload);
              } else {
                finalPayload = payload;
              }
            case 36:
              _context7.prev = 36;
              if (!(verb == 'POST')) {
                _context7.next = 43;
                break;
              }
              _context7.next = 40;
              return doFetch(url, {
                method: verb,
                body: finalPayload,
                headers: reqConfig.headers,
                responseType: config == null ? void 0 : config.responseType
              });
            case 40:
              resp = _context7.sent;
              _context7.next = 51;
              break;
            case 43:
              if (!(verb == 'GET')) {
                _context7.next = 50;
                break;
              }
              // TODO: Incorporate params into the query string
              if (payload) {
                url = url + '?' + new URLSearchParams(payload);
              }
              _context7.next = 47;
              return doFetch(url, {
                method: verb,
                headers: reqConfig.headers,
                responseType: config == null ? void 0 : config.responseType
              });
            case 47:
              resp = _context7.sent;
              _context7.next = 51;
              break;
            case 50:
              throw new SteamshipError({
                statusMessage: "Unsupported HTTP Verb: " + verb
              });
            case 51:
              _context7.next = 56;
              break;
            case 53:
              _context7.prev = 53;
              _context7.t1 = _context7["catch"](36);
              throw this._makeError({
                error: _context7.t1
              });
            case 56:
              return _context7.abrupt("return", this._makeResponse({
                response: resp,
                rawResponse: config == null ? void 0 : config.rawResponse,
                objectConstructor: config == null ? void 0 : config.expect,
                responsePath: config == null ? void 0 : config.responsePath
              }));
            case 57:
            case "end":
              return _context7.stop();
          }
        }, _callee7, this, [[36, 53]]);
      }));
      function call(_x14, _x15, _x16, _x17, _x18) {
        return _call.apply(this, arguments);
      }
      return call;
    }()
  }]);
  return ApiBase;
}();
